acvm/src/lib.rs

```diff
@@ -52,6 +52,8 @@ pub enum OpcodeResolutionError {
     UnexpectedOpcode(&'static str, BlackBoxFunc),
     #[error("expected {0} inputs for function {1}, but got {2}")]
     IncorrectNumFunctionArguments(usize, BlackBoxFunc, usize),
+    #[error("opcode solve failed: {0}")]
+    OpcodeSolveFailed(String),
```

Comment:
This name is too generic, perhaps it could be changed to something like:

```rust
    #[error("failed to solve blackbox function: {0}, reason: {1}")]
     BlackBoxFunctionFailed(BlackBoxFunc, String),
```

The reason for this is because a generic error type may then become the place where all new errors get placed. We ideally want to know exactly what the context is around the error that was being thrown, so that if we decide to remove blackbox function solving for example, we can then remove the accompanying error types and backends will know immediately what code needs to be revised.

Reply:
```suggestion
    #[error("failed to solve blackbox function: {0}, reason: {1}")] 
    BlackBoxFunctionFailed(BlackBoxFunc, String),
```
Yep, I'm fine with that; however, note that #257 will no longer pass the BlackBoxFunc to the backend but we can adjust the solve function to map the error to add the information